types/types.go

```diff
@@ -567,21 +566,68 @@ func (t Any) Merge(other Type) Any {
 
 // Union returns a new Any type that is the union of the two Any types.
 func (t Any) Union(other Any) Any {
+	// Return the more general (blank) Any type if present.
 	if len(t) == 0 {
 		return t
 	}
 	if len(other) == 0 {
 		return other
 	}
-	cpy := make(Any, len(t))
-	copy(cpy, t)
-	for i := range other {
-		if !cpy.Contains(other[i]) {
-			cpy = append(cpy, other[i])
+	// Prealloc the output list.
+	maxLen := len(t)
+	if len(t) < len(other) {
+		maxLen = len(other)
+	}
+	merged := make(Any, 0, maxLen)
+	// Note(philipc): Create a merged slice, doing the minimum number of
+	// comparisons along the way. We treat this as a problem of merging two
+	// sorted lists that might have duplicates. This specifically saves us
+	// from cases where one list might be *much* longer than the other.
+	// Algorithm:
+	//   Assume:
+	//   - List A
+	//   - List B
+	//   - List Output
+	//   - Idx_a, Idx_b
+	//   Procedure:
+	//   - While Idx_a < len(A) and Idx_b < len(B)
+	//     - Compare head(A) and head(B)
+	//     - Cases:
+	//       - A < B: Append head(A) to Output, advance Idx_a
+	//       - A == B: Append head(A) to Output, advance Idx_a, Idx_b
+	//       - A > B: Append head(B) to Output, advance Idx_b
+	//   - Return output
+	idxA := 0
+	idxB := 0
+	for idxA < len(t) || idxB < len(other) {
+		// Early-exit cases:
+		if idxA == len(t) {
+			// Ran out of elements in t. Copy over what's left from other.
+			merged = append(merged, other[idxB:]...)
+			break
+		} else if idxB == len(other) {
+			// Ran out of elements in other. Copy over what's left from t.
+			merged = append(merged, t[idxA:]...)
+			break
+		}
+		// Normal selection of next element to merge:
+		switch Compare(t[idxA], other[idxB]) {
+		// A < B:
+		case -1:
+			merged = append(merged, t[idxA])
+			idxA++
+		// A == B:
+		case 0:
+			merged = append(merged, t[idxA])
+			idxA++
+			idxB++
+		// A > B:
+		case 1:
+			merged = append(merged, other[idxB])
+			idxB++
 		}
 	}
-	sort.Sort(typeSlice(cpy))
-	return cpy
+	return merged
```

Comment:
We sort the final result in the original which we don't here.  Do we need that here as the final result may not be sorted or I'm probably missing something.

Reply:
The result is sorted *as it's constructed*.  I'm building off the assumption that the incoming `Any` slices *should* already be sorted, since both construction and `Merge` operations ensure sorted order of the slice. If one of the slices was not sorted already, the existing `Union` algorithm would also give erroneous results sometimes.

In the new algorithm, the `Compare` operations are done as the result-building loop iteratively walks along both (already sorted!) slices of `Type` objects, and the loop places the sorted-first item from one of those slices into the result slice. Counter variables keep track of which slice had an item taken from it for the result. When the loop runs out of items in one of the slices, it copies over all of the remaining (sorted) items from the remaining `Any` slice, because they would have sorted-after the last item in the merged result list anyway.

This should preserve the sort order perfectly, allowing us to skip the `sort.Sort` cleanup that used to be needed at the end of the function.

Reply:
> I'm building off the assumption that the incoming Any slices should already be sorted

We should add a note about this assumption in the method comment.

Reply:
I've added a note about it. Previously, it was just an undocumented assumption. :sweat_smile: